src/mapping/query/functions.rs

```diff
@@ -32,12 +119,16 @@ impl Function for NotFn {
 #[derive(Debug)]
 pub(in crate::mapping) struct ToStringFn {
     query: Box<dyn Function>,
-    default: Option<Value>,
+    default: Option<Box<dyn Function>>,
```

Comment:
_all_ arguments are now lazily evaluated.

This has both advantages and disadvantages:

advantages:

- all values can uniformly be provided by other event fields (e.g. `to_string(.nope, default = .foo)`)
- values are not computed unless required (e.g. `to_string(.foo, default = .nope)`)

disadvantage:

- value type validation has to happen at runtime
- some computational-heavy logic has to be pushed from compile to runtime (e.g. parsing timestamp string formats)

I think this is a good place to start, but we can optionally expand function parameters to expose if they accept computed values or not.

Alternatively, functions can try to downcase a `dyn Function` to a `Literal`, in which case the resulting literal can be used without the runtime context, which allows the function to pre-compute whatever complex logic it needs, to improve performance.

Reply:
I haven't done a before/after benchmark for the remap changes itself, but I have added a few more tests to compare remap vs regular transforms, and those look okay to me:

```
parse JSON with remap   time:   [1.4285 us 1.4439 us 1.4610 us]
parse JSON with json_parser                                                                             
                        time:   [1.4625 us 1.4788 us 1.4978 us]

coerce with remap       time:   [2.4033 us 2.4255 us 2.4500 us]
coerce with coercer     time:   [2.1415 us 2.1752 us 2.2134 us]
```

The `coerce with remap` benchmark parses the timestamp format string at runtime right now, whereas before it was done at compile-time, so there might be some overhead there that causes it to be ~0.3us slower than the regular coercer, but I'd call that good enough for now.

We can use a separate PR to start playing with optimizing functions by trying to downcast fields and store concrete fields in the function struct if the downcast worked.